### PR TITLE
Add IonQ native gateset support

### PR DESCRIPTION
**Pull request by Cynocracy**

### CODE REVIEW

**Comment by Cynocracy** [comment]
Will have to go to an older black version for formatting, I guess.

The TODOs need to be resolved before submission, and the too-few-public-method check should be skipped/ignored somehow

Can add tests once we finalize API shape

**Comment by josh146** [comment]
Thanks @Cynocracy!

**Comment by Cynocracy** [comment]
Ready for review!

**Review by ColemanCollins — APPROVED** [review]

File: pennylane_ionq/ops.py [hunk 1]
```diff
@@ -16,6 +16,68 @@
...
+                0 & e^{-i\phi} \\
+                e^{-i\phi} & 0
+            \end{pmatrix}
+    Args:
+        phi (float): phase :math:`\phi`
+        wires (Sequence[int]): the subsystems the operation acts on
+    """
+    num_params = 1
+    num_wires = 1
+    par_domain = "R"
```
Comment (albi3ro) [inline pennylane_ionq/ops.py:35]: I believe `par_domain` is now a deprecated attribute, so we do not need to specify this anymore.
```suggestion
```

Comment (Cynocracy) [reply pennylane_ionq/ops.py:35]: Ah, cool!

**Review by albi3ro — CHANGES_REQUESTED** [review]
I was able to submit a job to the simulator with the native gate set, but `"ionq.qpu"` is hanging currently for both "native" and "qis" gates.

Exciting to be able to work with these gates!

Most of my notes have to do with documentation.

We could probably add parameter-shift support later for these gates, but I think it's ok to leave them with `grad_method = None` for now.

File: pennylane_ionq/__init__.py [hunk 2]
```diff
@@ -15,5 +15,6 @@
 PennyLane IonQ overview
 =======================
 """
+from .ops import GPI, GPI2, MS
```
[the remark this answers is not in the document]
Comment (Cynocracy) [reply pennylane_ionq/__init__.py:18]: We can do that! I'm not sure if they ended up being used

File: pennylane_ionq/ops.py [hunk 3]
```diff
@@ -16,6 +16,66 @@
...
+                0 & e^{-i*2*\pi*\phi} \\
+                e^{-i*2*\pi*\phi} & 0
+            \end{pmatrix}
+    Args:
+        phi (float): phase :math:`\phi`
+        wires (Sequence[int]): the subsystems the operation acts on
+    """
+    num_params = 1
+    num_wires = 1
+    grad_method = "None"
```
[the remark this answers is not in the document]
Comment (Cynocracy) [reply pennylane_ionq/ops.py:35]: Do you have an example of how to add parameter shift support? I'd be happy to give it a shot (though maybe in a follow-up PR)

File: pennylane_ionq/ops.py [hunk 4]
```diff
@@ -16,6 +16,66 @@
 """
 from pennylane.operation import Operation
 
+# Custom operations for the native gateset below.
+class GPI(Operation):  # pylint: disable=too-few-public-methods
+    r"""GPI(phi, wires)
+    Single-qubit GPI gate.
+    .. math::
+        GPI(\phi) =
```
[the remark this answers is not in the document]
Comment (Cynocracy) [reply pennylane_ionq/ops.py:24]: :+1: Thanks! will check

**Comment by Cynocracy** [comment]
Is the QPU job submitting fine but then not being processed? We have very long queue times at the moment: https://status.ionq.co/

**Comment by Cynocracy** [comment]
Okay, after many mini fixups I think the docs render and don't have any glaring typos (I hope!)

File: pennylane_ionq/ops.py [hunk 5]
```diff
@@ -16,6 +16,72 @@
...
+    r"""GPI(phi, wires)
+    Single-qubit GPI gate.
+
+    .. math::
+
+       GPI(\phi) =
+            \begin{pmatrix}
+                0 & e^{-i*2*\pi*\phi} \\
+                e^{i*2*\pi*\phi} & 0
+            \end{pmatrix}
```
Comment (albi3ro) [inline pennylane_ionq/ops.py:30]:
```suggestion
            \begin{pmatrix}
                0 & e^{-i 2 \pi \phi} \\
                e^{i 2 \pi \phi} & 0
            \end{pmatrix}
```

File: pennylane_ionq/ops.py [hunk 6]
```diff
@@ -16,6 +16,72 @@
...
+    r"""GPI2(phi, wires)
+    Single-qubit GPI2 gate.
+
+    .. math::
+
+       GPI2(\phi) =
+            \begin{pmatrix}
+                 1 & -i*e^{-i*2*\pi*\phi} \\
+                 -i*e^{i*2*\pi*\phi} & 1
+            \end{pmatrix}
```
Comment (albi3ro) [inline pennylane_ionq/ops.py:50]:
```suggestion
            \begin{pmatrix}
                 1 & -i e^{-2 \pi i \phi} \\
                 -i e^{2 \pi i \phi} & 1
            \end{pmatrix}
```

File: pennylane_ionq/ops.py [hunk 7]
```diff
@@ -16,6 +16,72 @@
...
+    2-qubit entanlging MS gate.
+
+    .. math::
+
+       MS(\phi_{0}, \phi_{1}) =
+            \frac{1}{\sqrt{2}}\begin{pmatrix}
+                1 & 0 & 0 & -i*e^{-i*2*\pi*(\phi_{0}+\phi_{1})} \\
+                0 & 1 & -i*e^{-i*2*\pi*(\phi_{0}-\phi_{1})} & 0 \\
+                0 & -i*e^{i*2*\pi*(\phi_{0}-\phi_{1})} & 1 & 0 \\
+                -i*e^{i*2*\pi*(\phi_{0}+\phi_{1})} & 0 & 0 & 1
```
Comment (albi3ro) [inline pennylane_ionq/ops.py:71]:
```suggestion
                1 & 0 & 0 & -i e^{-2 \pi i(\phi_{0}+\phi_{1})} \\
                0 & 1 & -i e^{-2 \pi i (\phi_{0}-\phi_{1})} & 0 \\
                0 & -i e^{2 \pi i(\phi_{0}-\phi_{1})} & 1 & 0 \\
                -i e^{2 \pi i(\phi_{0}+\phi_{1})} & 0 & 0 & 1
```

**Review by albi3ro — COMMENTED** [review]
The `.. math::` accepts LaTeX, so there's no need for the `*` symbols.

Thanks for updating the docstring for the `IonQDevice` class. We should similarly update the docstring for the `QPUDevice` and the `SimulatorDevice`.

File: pennylane_ionq/ops.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane_ionq/ops.py]: The easiest way is to define the [`parameter_frequencies`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.operation.Operation.html#pennylane.operation.Operation.parameter_frequencies) property. 

If the parameter frequencies aren't manually defined, we can calculate them from the `generator`, defined via the `generator` method.

As these are single qubit operators with a single parameter, they will probably have the same frequencies as the standard rotation gates, namely `[(1,)]`, but I'd have to double check.

**Comment by Cynocracy** [comment]
docs look good!

**Review by albi3ro — APPROVED** [review]
Looks great! Thanks for being patient will all the changes.

**Comment by Cynocracy** [comment]
No problem :) I appreciate the thorough review!

**Pull request merged**